### PR TITLE
Style heading pattern's typography

### PR DESCRIPTION
**Pull request by davidcmoulton**
No web fonts yet.

### CODE REVIEW

File: src/core-styling/functions/_validation.scss [hunk 1]
```diff
@@ -0,0 +1,9 @@
+$_is-test: false !default;
+
+@function error($message, $capture: $_is-test) {
```
Comment (davidcmoulton) [inline src/core-styling/functions/_validation.scss]: Previously was `_error`, but that enforces a privacy when `@use`d that we don't want in this case.

File: src/core-styling/functions/_validation.scss [hunk 2]
```diff
@@ -0,0 +1,9 @@
+$_is-test: false !default;
```
Comment (davidcmoulton) [inline src/core-styling/functions/_validation.scss]: Should not be private.

File: src/core-styling/mixins/_typography.scss [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+@use "types";
+@use "utilities";
+
+@mixin set-font-size-and-line-height($font-size, $block-size: baselinegrid.$space-small) {
+  @include types.rem(font-size, $font-size);
+  line-height: $block-size / $font-size;
+}
+
+@mixin covert-link($hover-color: color.$primary-dark) {
+  @include utilities.inherit-all($ensure: color text-decoration) {
```
Comment (davidcmoulton) [inline src/core-styling/mixins/_typography.scss]: Still not sure whether `inherit-all` is too blunt an instrument. Will leave it for now, to reconsider later.

File: src/patterns/molecules/heading/heading.scss [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+@mixin _typography($heading-level) {
+  $permitted: (level1, level2, level3, level4, level5, level6);
+  @if not list.index($permitted, $heading-level) {
+    @include validation.error("Invalid heading level supplied: #{$heading-level}");
+  }
+
+  $scale: _get-scale-point($heading-level);
+  $line-height: _get-line-height($heading-level);
+  @include _base-typography();
+  @include typography.set-font-size-and-line-height(scale($scale), $line-height);
```
Comment (davidcmoulton) [inline src/patterns/molecules/heading/heading.scss]: Missing `scale`.

File: src/core-styling/mixins/_scale.scss [hunk 5]
```diff
@@ -0,0 +1,6 @@
+@use "../settings/font";
+@use "../../../node_modules/modularscale-sass/stylesheets/modularscale";
```
Comment (davidcmoulton) [inline src/core-styling/mixins/_scale.scss]: Not that happy about reaching down directly into `node_modules`, but equally not happy about tying the sass implementation to Webpack by [using the `~` prefix](https://webpack.js.org/loaders/sass-loader/#resolving-import-at-rules) either 🤔.

File: src/patterns/molecules/heading/heading.scss [hunk 6]
```diff
@@ -0,0 +1,82 @@
+@use "sass:list";
+@use "sass:map";
+@use "../../../core-styling/settings/baselinegrid";
+@use "../../../core-styling/settings/font";
+@use "../../../core-styling/mixins/typography";
+@use "../../../core-styling/mixins/scale";
+@use "../../../core-styling/mixins/validation";
+
+$_h1-line-height: baselinegrid.$space-medium;
```
Comment (davidcmoulton) [inline src/patterns/molecules/heading/heading.scss]: Alternative would be to make these configurable with an implementation like e.g.: 
```
$h1-line-height: baselinegrid.$space-medium !default;
```
However this is outside the scope of configurability we're allowing, so unless there're good reasons not to, will keep them as private, and soley defined by the baseline grid.

File: src/core-styling/mixins/_types.scss [hunk 7]
```diff
@@ -0,0 +1,44 @@
+@use "../settings/font";
+@use "../../../node_modules/sass-rem/rem";
+
+$rem-baseline: font.$size;
+$rem-fallback: true;
```
Comment (davidcmoulton) [inline src/core-styling/mixins/_types.scss]: Can safely be `false` now.

Comment (thewilkybarkid) [reply src/core-styling/mixins/_types.scss]: Would simplify things, and means we don't need the override either.

File: src/core-styling/mixins/_scale.scss (hunk 5 above)
Comment (thewilkybarkid) [reply src/core-styling/mixins/_scale.scss]: Since the path won't work if this were turned into an NPM module, probably better to use the Webpack syntax for now...

File: src/patterns/molecules/heading/heading.scss [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+@mixin _base-typography {
+  font-family: $_font-family;
+  font-weight: 600;
+}
+
+@mixin _typography($heading-level) {
+  $permitted: (level1, level2, level3, level4, level5, level6);
+  @if not list.index($permitted, $heading-level) {
+    @include validation.error("Invalid heading level supplied: #{$heading-level}");
```
Comment (thewilkybarkid) [inline src/patterns/molecules/heading/heading.scss]: Since this is private, this isn't needed (could never test for it).

File: src/core-styling/mixins/_scale.scss (hunk 5 above)
Comment (davidcmoulton) [reply src/core-styling/mixins/_scale.scss]: True enough 👍

File: src/patterns/molecules/heading/heading.scss (hunk 8 above)
Comment (davidcmoulton) [reply src/patterns/molecules/heading/heading.scss]: Yeah. I was getting carried away with _validateAllTheInputz!_

File: src/core-styling/settings/_font.scss [hunk 9]
```diff
@@ -0,0 +1,5 @@
+$letterspacing-label: 0.5px !default;
```
Comment (thewilkybarkid) [inline src/core-styling/settings/_font.scss]: Not needed

File: src/core-styling/settings/_color.scss [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+$text-ui_background_hue: rgb(245, 245, 245) !default;
+$text-ui_code: rgb(247, 247, 247) !default;
+$text-ui_background__reverse: rgb(33, 33, 33) !default;
+$text-ui_background_hue__reverse: rgb(51, 51, 51) !default;
+$background: rgb(255, 255, 255) !default;
+$information: rgb(2, 136, 209) !default;
+$success: rgb(98, 159, 67) !default;
+$success_dark: rgb(86, 144, 55) !default;
+$attention: rgb(207, 12, 78) !default;
+$warning: rgb(230, 81, 0) !default;
```
Comment (thewilkybarkid) [inline src/core-styling/settings/_color.scss]: I recommend removing all the colours not needed in favour of some rethinking (I wouldn't be able to tell you what `text-ui_background_hue__reverse` means).

File: src/core-styling/mixins/_scale.scss [hunk 11]
```diff
@@ -0,0 +1,6 @@
...
+
+@function scale($value: 0) {
+  @return round(modularscale.ms-function($value, font.$size, modularscale.$major-second));
```
Comment (thewilkybarkid) [inline src/core-styling/mixins/_scale.scss]: I would recommend defining the base size and the scale together (one can be changed, the other can't).

File: src/patterns/molecules/heading/heading.scss [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  $scale: _get-scale-point($heading-level);
+  $line-height: _get-line-height($heading-level);
+  @include _base-typography();
+  @include typography.set-font-size-and-line-height(scale.scale($scale), $line-height);
+}
```
Comment (thewilkybarkid) [inline src/patterns/molecules/heading/heading.scss]: There's quite a lot of indirection here, would be easier to understand with a simpler implementation.

File: src/core-styling/mixins/_typography.scss (hunk 3 above)
Comment (thewilkybarkid) [reply src/core-styling/mixins/_typography.scss]: I think it proved to be, and should probably not carry it over.

File: src/core-styling/mixins/_scale.scss (hunk 11 above)
Comment (davidcmoulton) [reply src/core-styling/mixins/_scale.scss]: Yup. I'm working on an improvement around baseline and scale, based on the concepts we discussed yesterday.

File: src/core-styling/settings/_color.scss (hunk 10 above)
Comment (davidcmoulton) [reply src/core-styling/settings/_color.scss]: Agree. The configuration requirements, combined with the new flexibility provided by Sass modules, provide a good opportunity.

File: src/core-styling/settings/_baseline.scss [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+// It is not recommended that you modify these following variables (it can break your vertical rhythm).
+
+$unit-extra-small: $_baseline-unit * $_factor-extra-small;
+$unit-small: $_baseline-unit * $_factor-small;
+$unit-smallish: $_baseline-unit * $_factor-smallish;
+$unit-medium: $_baseline-unit * $_factor-medium;
+$unit-large: $_baseline-unit * $_factor-large;
+$unit-extra-large: $_baseline-unit * $_factor-extra-large;
+
+@function scale($value: 0) {
```
Comment (thewilkybarkid) [inline src/core-styling/settings/_baseline.scss]: Thinking about the new modules, probably doesn't make sense to split functions, mixins, settings and utilities anymore. (So we'd have cases like this, with settings, functions etc.)

File: src/core-styling/settings/_baseline.scss [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+$_factor-small: 1; // 24px
+$_factor-smallish: 1.5; // 36px
+$_factor-medium: 2; // 48px
+$_factor-large: 3; // 72px
+$_factor-extra-large: 5; // 120px
+
+//----------------------------------------------------------------------------------------------------------------------
+// Spacing values are determined based on the global baseline grid.
+//
+// It is not recommended that you modify these following variables (it can break your vertical rhythm).
```
Comment (thewilkybarkid) [inline src/core-styling/settings/_baseline.scss]: If there's no `!default`, they can't be modified can they?

File: src/core-styling/settings/_baseline.scss [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+$unit-extra-small: $_baseline-unit * $_factor-extra-small;
+$unit-small: $_baseline-unit * $_factor-small;
+$unit-smallish: $_baseline-unit * $_factor-smallish;
```
Comment (thewilkybarkid) [inline src/core-styling/settings/_baseline.scss]: Time to revisit the names? 😅

File: src/patterns/molecules/heading/heading.scss [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+
+@mixin _typography($heading-level) {
+  $scale: _get-scale-point($heading-level);
+  $line-height: _get-line-height($heading-level);
+  @include _base-typography();
+  @include typography.set-font-size-and-line-height(baseline.scale($scale), $line-height);
+}
+
+h1 {
```
Comment (thewilkybarkid) [inline src/patterns/molecules/heading/heading.scss]: Bit of a different implementation, where the basic styling was done in a global normalization (so you didn't _have_ to use the heading pattern to have a `<h1>` etc).

Think it does make sense to try and do this, but makes we wonder about normalization. (Would we do the same with lists, for example?)

File: src/patterns/molecules/heading/heading.scss (hunk 12 above)
Comment (davidcmoulton) [reply src/patterns/molecules/heading/heading.scss]: Hopefully https://github.com/libero/storybook/pull/21/commits/6bf05199374456029a484486c0063036d13a8e11 makes it clearer 😅

File: src/core-styling/settings/_baseline.scss (hunk 15 above)
Comment (davidcmoulton) [reply src/core-styling/settings/_baseline.scss]: 😅

File: src/core-styling/settings/_baseline.scss (hunk 13 above)
Comment (davidcmoulton) [reply src/core-styling/settings/_baseline.scss]: I think so. Will need to be mindful of not blocking configuration where necessary, but now we're not looking to share cross-product sass variables in the way we previously were, it gives more room to manoeuvre here.

File: src/patterns/molecules/heading/heading.scss (hunk 16 above)
Comment (davidcmoulton) [reply src/patterns/molecules/heading/heading.scss]: I'm not sure yet that this is the final implementation. Going for encapsulation first and then to break things out where it makes sense to, hopeful to be able to discover a happy medium.

Comment (davidcmoulton) [reply src/patterns/molecules/heading/heading.scss]: I'm going to try breaking the heading typography into its own module. It can then be `@use`d in both the heading and normalisation stylesheets. A couple of other things I want to try too...

File: .storybook/webpack.config.js [hunk 17]
```diff
@@ -1,17 +1,19 @@
...
+            autoescape: true,
+            strict_variables: true,
+          },
         },
-      },
-    ],
-  });
+      ],
+    },
+  );
```
Comment (thewilkybarkid) [inline .storybook/webpack.config.js]: Looks like just a CS change?

File: src/core-styling/_baseline.scss [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+$_factor-extra-small: 0.5; // 12px
+$_factor-small: 1; // 24px
+$_factor-smallish: 1.5; // 36px
+$_factor-medium: 2; // 48px
+$_factor-large: 3; // 72px
+$_factor-extra-large: 5; // 120px
+
+//----------------------------------------------------------------------------------------------------------------------
+// Spacing values are determined based on the global baseline grid.
+//
```
Comment (thewilkybarkid) [inline src/core-styling/_baseline.scss]: Might want to be consistent on how to comment. 😅

File: src/core-styling/_types.scss [hunk 19]
```diff
@@ -0,0 +1,7 @@
+@use "settings/font";
+@use "~sass-rem/rem" with (
+  $rem-baseline: font.$size,
+  $rem-fallback: false
+);
+
+@forward "~sass-rem/rem";
```
Comment (thewilkybarkid) [inline src/core-styling/_types.scss]: This module should be called `rem`?

File: src/core-styling/_baseline.scss [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+
+$unit-extra-small: $_baseline-unit * $_factor-extra-small;
+$unit-small: $_baseline-unit * $_factor-small;
+$unit-smallish: $_baseline-unit * $_factor-smallish;
+$unit-medium: $_baseline-unit * $_factor-medium;
+$unit-large: $_baseline-unit * $_factor-large;
+$unit-extra-large: $_baseline-unit * $_factor-extra-large;
+
+@function scale($value: 0) {
```
Comment (thewilkybarkid) [inline src/core-styling/_baseline.scss]: Not sure this should have a default?

File: src/core-styling/_link.scss [hunk 21]
```diff
@@ -0,0 +1,12 @@
+@use "settings/color";
+
+@mixin covert-link($hover-color: color.$primary-dark) {
+  color: inherit;
+  cursor: pointer;
+  display: inline;
+  text-decoration: none;
+
+  &:hover {
+    color: color.$primary-dark;
```
Comment (thewilkybarkid) [inline src/core-styling/_link.scss]:
```suggestion
    color: $hover-color;
```

File: src/core-styling/_normalize.scss [hunk 22]
```diff
@@ -0,0 +1,6 @@
+*,
+*:before,
+*:after {
+  box-sizing: border-box;
+}
+
```
Comment (thewilkybarkid) [inline src/core-styling/_normalize.scss]:
```suggestion
```

File: src/core-styling/_typography-heading.scss [hunk 23]
```diff
@@ -0,0 +1,67 @@
+@forward "typography-global";
```
Comment (thewilkybarkid) [inline src/core-styling/_typography-heading.scss]: Don't think it should?

File: src/core-styling/_typography-heading.scss [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+  @include _typography(1, $_default_line-height);
+}
+
+@mixin h6-typography() {
+  @include _typography(0, $_default_line-height);
+}
+
+@mixin _typography($scale-point, $line-height) {
+  font-family: $_font-family;
+  font-weight: 600;
```
Comment (thewilkybarkid) [inline src/core-styling/_typography-heading.scss]: Might as well make that a variable too.

File: src/core-styling/_typography-heading.scss [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+
+@mixin h5-typography() {
+  @include _typography(1, $_default_line-height);
+}
+
+@mixin h6-typography() {
+  @include _typography(0, $_default_line-height);
+}
+
+@mixin _typography($scale-point, $line-height) {
```
Comment (thewilkybarkid) [inline src/core-styling/_typography-heading.scss]: Wonder about the order of the file (public vs private API, and selectors).

File: src/core-styling/_types.scss (hunk 19 above)
Comment (thewilkybarkid) [inline src/core-styling/_types.scss]:
```suggestion
@forward "~sass-rem/rem" show rem;
```

(Not tested; does this forward both the mixin and the function?)

File: src/core-styling/_typography-heading.scss [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+@use "settings/font";
+@use "baseline";
+@use "typography-global" as global-typography;
+
+$_h1-line-height: baseline.$unit-medium;
+$_h2-line-height: (baseline.$unit-small + baseline.$unit-smallish) / 2;
+$_default_line-height: baseline.$unit-small;
+$_font-family: font.$secondary;
+
+@mixin h1-typography() {
```
Comment (thewilkybarkid) [inline src/core-styling/_typography-heading.scss]: Wondering whether this should just be `h1`? The module is specifically about typography, so `heading-typography.h1-typography` seems a bit redundant.

File: src/core-styling/_typography-heading.scss (hunk 23 above)
Comment (davidcmoulton) [reply src/core-styling/_typography-heading.scss]: 👍 Old code. Removing.

File: src/core-styling/_typography-heading.scss (hunk 25 above)
Comment (davidcmoulton) [reply src/core-styling/_typography-heading.scss]: Order matters for compilation: the mixins need to be defined before they're used in the selectors, and the local variables need to be defined before the mixin they're called from is executed.

Update: have changed the order as much as practicable within constraints.

File: src/core-styling/_types.scss (hunk 19 above)
Comment (davidcmoulton) [reply src/core-styling/_types.scss]: Don't like the ambiguity there. `@forward` removed in the refactor, see https://github.com/libero/storybook/pull/21/files#diff-cfeef31a60b46e4584db50a2f02af416

File: src/core-styling/_baseline.scss [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+$_factor-small: 1; // 24px
+$_factor-smallish: 1.5; // 36px
+$_factor-medium: 2; // 48px
+$_factor-large: 3; // 72px
+$_factor-extra-large: 5; // 120px
+
+//----------------------------------------------------------------------------------------------------------------------
+// Spacing values are determined based on the global baseline grid.
+//
+$unit-extra-small: $_baseline-unit * $_factor-extra-small;
```
Comment (davidcmoulton) [inline src/core-styling/_baseline.scss]: (Will supply a followup PR is to move to a 12px baseline unit, and replace the `$unit-[x]` variables with a function.)

File: src/core-styling/_scale.scss [hunk 28]
```diff
@@ -0,0 +1,10 @@
+@use "sass:math";
+@use "font";
+@use "~modularscale-sass/stylesheets/modularscale";
+
+$base: font.$size;
+$ratio: modularscale.$major-second !default;
+
+@function at($point) {
```
Comment (thewilkybarkid) [inline src/core-styling/_scale.scss]: Might need a better name.

**Review by thewilkybarkid — APPROVED** [review]
I'm made quite a few changes re the modularisation. Hopefully for the better, but please took take a look and revert if they don't make sense.

File: src/shared-styles/_baseline.scss [hunk 29]
```diff
@@ -0,0 +1,38 @@
...
+$_baseline-unit: scale.$base * 0.375; // 6px
+
+//----------------------------------------------------------------------------------------------------------------------
+// Spacing values determined by the global baseline grid.
+//
+@function _baseline-unit($factor) {
+  @return $_baseline-unit * $factor;
+}
+
+$size-1: _baseline-unit(1); // 6px
```
Comment (davidcmoulton) [inline src/shared-styles/_baseline.scss]: Could make these private and expose those currently required via names of convenience. It raises the question of what those names should be. Some options are:
1. keep `size` variables public and just use these. Not a bad option, and it means we don't have to keep coming up with sensible convenience names whenever a new `size` value needs making public
1. make `size` variables private and use the original names for convenience public variables or
1.  make `size` variables private and use newly suggested names for convenience public variables or
1.  make `size` variables private and use Reviewer's names for convenience public variables if they would work sensibly here too.

Right now options 1 and 4 look the best, with a slight preference for 1.

Comment (davidcmoulton) [reply src/shared-styles/_baseline.scss]: **Update**: Decided to use a [function](https://github.com/libero/storybook/pull/21/files#diff-cec04b6d5a3012af00cfbdfa5a559686R15) to get the appropriate from the number of baseline units, comparable to the [`scale.at()` function](https://github.com/libero/storybook/pull/21/files#diff-e57318c7e4db838edcbdc32fb2979662R8)

File: src/shared-styles/_baseline.scss [hunk 30]
```diff
@@ -12,27 +12,27 @@ $_baseline-unit: scale.$base * 0.375; // 6px
...
+$_size-11: calculateSizeFromUnits(11); // 66px
+$_size-12: calculateSizeFromUnits(12); // 72px
+$_size-13: calculateSizeFromUnits(13); // 78px
+$_size-14: calculateSizeFromUnits(14); // 84px
+$_size-15: calculateSizeFromUnits(15); // 90px
+$_size-16: calculateSizeFromUnits(16); // 96px
+$_size-17: calculateSizeFromUnits(17); // 102px
+$_size-18: calculateSizeFromUnits(18); // 108px
+$_size-19: calculateSizeFromUnits(19); // 114px
+$_size-20: calculateSizeFromUnits(20); // 120px
```
Comment (thewilkybarkid) [inline src/shared-styles/_baseline.scss]: Aren't needed anymore?

File: src/shared-styles/_baseline.scss [hunk 31]
```diff
@@ -0,0 +1,38 @@
...
+//
+// Every spacing metric should be based on this.
+// All commented sizes assume a scale base of 16px.
+//
+$_baseline-unit: scale.$base * 0.375; // 6px
+
+//----------------------------------------------------------------------------------------------------------------------
+// Spacing values determined by the global baseline grid.
+//
+@function calculate-size-from-units($factor) {
```
Comment (thewilkybarkid) [inline src/shared-styles/_baseline.scss]: Hmm, wordy. 'calculate' is probably redundant.

File: src/shared-styles/_baseline.scss [hunk 32]
```diff
@@ -0,0 +1,18 @@
...
+//
+$_baseline-unit: scale.$base * 0.375; // 6px
+
+//----------------------------------------------------------------------------------------------------------------------
+// Spacing values determined by the global baseline grid.
+//
+@function size-from-units($factor) {
+  @return $_baseline-unit * $factor;
+}
+
```
Comment (thewilkybarkid) [inline src/shared-styles/_baseline.scss]:
```suggestion
```

**Review by thewilkybarkid — APPROVED** [review]
Couple of naming tweaks (😅), but think we're good to :shipit:.
